samples/99_zil_interpreter/app/main.rb

```diff
@@ -7,39 +7,108 @@
 
 def tick(args)
   setup(args) if args.tick_count.zero?
-  # TODO:
-  # - Render history
-  # - Collect input
-  # - Call send_input with input when pressing Enter
+  # Render history
+  render_state(args)
+  
+  # Collect input
+  handle_input(args)
+  
+  # Call send_input when pressing Enter
+  if args.inputs.keyboard.key_down.enter
+    send_input(args, args.state.input)
+  end
+  
   $gtk.request_quit unless $interpreter.alive?
 end
 
 def setup(args)
+  args.state.text_color = [240, 255, 255]
+  args.state.bg_color = [0, 0, 0]
+  args.state.input = ''
+  args.state.text_history = []
+  args.state.current_line = 0
   context = build_zil_context(args)
   args.state.zil_context = context
   $interpreter = Fiber.new {
     context.globals[:GO].call [], context
   }
   $interpreter.resume # Initial processing until first Fiber.yield
-  process_outputs args, context.outputs # Process welcome message if existing
+  process_outputs(args, context.outputs) # Process welcome message if existing
 
   # TODO:
   # Add other setup if necessary
 end
 
-# Call this method with the input after pressing enter
+def handle_input(args)
+  if args.inputs.keyboard.key_down.backspace
+    args.state.input.chop!
+  else
+    if args.inputs.text[0]
+      args.state.input << args.inputs.text[0]
+      args.state.current_line = 0
+    end
+  end
+
+  # Scrolling
+  if args.inputs.mouse.wheel
+    args.state.current_line += args.inputs.mouse.wheel.y
+
+    # Stop the player from scrolling past the beginning or end
+    max_possible_line = [0, args.state.text_history.length - 33].max
+    args.state.current_line = args.state.current_line.clamp(0, max_possible_line)
+
+  # Jump to the present
+  elsif args.inputs.keyboard.key_down.escape
+    args.state.current_line = 0
+  end
+end
+
+# Called with the input after pressing enter
 def send_input(args, input)
+  return if input.empty?
+
   $interpreter.resume input
 
+  args.state.text_history << " "
+  args.state.text_history << "> #{input}"
+  args.state.input = ''
+  args.state.current_line = 0
+
   context = args.state.zil_context
   process_outputs(args, context.outputs)
-  context.outputs.clear
 end
 
 def process_outputs(args, outputs)
-  # TODO:
-  # Replace with actual processing, e.g. adding it to the history of output lines etc
-  outputs.each do |line|
-    puts line
+  args.state.text_history += outputs
+  args.state.zil_context.outputs.clear
+end
+
+def render_state(args)
+  args.outputs.background_color = args.state.bg_color
+  
+  # Player input
+  input_line = "> #{args.state.input}"
+  input_line << "_" if (args.tick_count / 32).round.mod_zero? 2 # Blinky underscore like all the old computers!
+  args.outputs.labels << {
+    x: 2, y: 22,
+    text: input_line,
+    r: args.state.text_color[0],
+    g: args.state.text_color[1],
+    b: args.state.text_color[2]
+  }
+  
+  # Computer response
+  text_start = args.state.text_history.length - args.state.current_line
+  line = 0
+  while line < 34
```

Comment:
Let's extract this into a constant after all - and change the conditional to a `<=`
```suggestion
  while line <= MAX_DISPLAYED_LINES
```

Reply:
Then the 33 in  the `max_possible_offset` above can also be replaced with the same constant